operator/pkg/bgp/manager.go

```diff
@@ -0,0 +1,481 @@
+			for e := range b.bgpNCOResource.Events(ctx) {
+				b.triggerReconcile()
+				e.Done(nil)
+			}
+			return nil
+		}),
+
+		job.OneShot("Node tracker", func(ctx context.Context, health cell.HealthReporter) error {
+			for e := range b.nodeResource.Events(ctx) {
+				b.triggerReconcile()
+				e.Done(nil)
+			}
+			return nil
+		}),
+	)
+
+	b.lc.Append(jobGroup)
+}
+
+func (b *BGPResourceManager) initializeStores(ctx context.Context) (err error) {
+	b.bgpCCStore, err = b.bgpCCResource.Store(ctx)
+	if err != nil {
+		return
+	}
+
+	b.bgpNCOStore, err = b.bgpNCOResource.Store(ctx)
+	if err != nil {
+		return
+	}
+
+	b.bgpNCStore, err = b.bgpNCResource.Store(ctx)
+	if err != nil {
+		return
+	}
+
+	b.bgpAStore, err = b.bgpAResource.Store(ctx)
+	if err != nil {
+		return
+	}
+
+	b.bgpPCStore, err = b.bgpPCResource.Store(ctx)
+	if err != nil {
+		return
+	}
+
+	b.nodeStore, err = b.nodeResource.Store(ctx)
+	if err != nil {
+		return
+	}
+
+	b.bgpPPStore, err = b.bgpPPResource.Store(ctx)
+	if err != nil {
+		return
+	}
+
+	return nil
+}
+
+// triggerReconcile initiates level triggered reconciliation.
+func (b *BGPResourceManager) triggerReconcile() {
+	select {
+	case b.reconcileCh <- struct{}{}:
+		b.logger.Debug("BGP reconciliation triggered")
+	default:
+	}
+}
+
+// Run starts the BGPResourceManager operator.
+func (b *BGPResourceManager) Run(ctx context.Context) (err error) {
+
+	// make sure both policy and cluster config are synced before starting the reconciliation
+	<-b.bgpClusterSyncCh
+	<-b.bgpPolicySyncCh
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+
+		case _, open := <-b.reconcileCh:
+			if !open {
+				return
+			}
+
+			err := b.reconcileWithRetry(ctx)
+			if err != nil {
+				b.logger.WithError(err).Error("BGP reconciliation failed")
+			} else {
+				b.logger.Debug("BGP reconciliation successful")
+			}
+		}
+	}
+}
+
+// reconcileWithRetry retries reconcile with exponential backoff.
+func (b *BGPResourceManager) reconcileWithRetry(ctx context.Context) error {
+	retryFn := func(ctx context.Context, retries int) (bool, error) {
+		err := b.reconcile(ctx)
+
+		switch {
+		case err != nil && errors.Is(err, ErrClusterAndPeeringPolicyEnabled):
+			// unrecoverable error, stop retry
+			return false, err
+		case err != nil:
+			// log error, continue retry
+			b.logger.WithError(TrimError(err, maxErrorLen)).Warnf("BGP reconciliation error, retry %d/%d", retries, maxRetries)
+			return false, err
+		default:
+			// no error, stop retry
+			return true, nil
+		}
+	}
+
+	return resiliency.Retry(ctx, initRetryDuration, maxRetries, retryFn)
+}
+
+// reconcile is called when any interesting resource change event is triggered.
+// We do not reconcile if both BGP Peering policy and BGP Cluster Config are enabled.
+func (b *BGPResourceManager) reconcile(ctx context.Context) error {
+	var err error
+	ppEnabled, ccEnabled := len(b.bgpPPStore.List()) > 0, len(b.bgpCCStore.List()) > 0
```

Comment:
Okay so for now, if both cilium bgp peering policy and cluster config are present - cluster config will get configured. 

I need to think a little bit more of the corner cases and add some tests which showcase the migration. Potentially a follow up PR on this topic.

Reply:
Yeah, I'm also not certain if it is ok to just prioritize ClusterConfig. In that case, maybe we cannot roll back when something goes wrong during migration? I think it's worth having an upgrade plan on the CFP. If we can have an automated upgrade/downgrade test, that would be the best.

However, I don't think we need to block this PR because of that. We won't make v2 API generally available, so if we make it clear that the behavior is subject to be changed, then I think it's ok to merge this as is.